osu.Game.Rulesets.Osu/UI/OsuResumeOverlay.cs

```diff
@@ -35,23 +36,26 @@ private void load()
         {
             OsuResumeOverlayInputBlocker? inputBlocker = null;
 
-            if (drawableRuleset != null)
+            var drawableOsuRuleset = (DrawableOsuRuleset?)drawableRuleset;
+
+            if (drawableOsuRuleset != null)
             {
-                var osuPlayfield = (OsuPlayfield)drawableRuleset.Playfield;
+                var osuPlayfield = drawableOsuRuleset.Playfield;
                 osuPlayfield.AttachResumeOverlayInputBlocker(inputBlocker = new OsuResumeOverlayInputBlocker());
             }
 
             Add(cursorScaleContainer = new Container
             {
                 Child = clickToResumeCursor = new OsuClickToResumeCursor
                 {
-                    ResumeRequested = () =>
+                    ResumeRequested = action =>
                     {
                         // since the user had to press a button to tap the resume cursor,
                         // block that press event from potentially reaching a hit circle that's behind the cursor.
                         // we cannot do this from OsuClickToResumeCursor directly since we're in a different input manager tree than the gameplay one,
                         // so we rely on a dedicated input blocking component that's implanted in there to do that for us.
-                        if (inputBlocker != null)
+                        // note this only matters when the user didn't pause while they were holding the same key that they are resuming with.
+                        if (inputBlocker != null && !drawableOsuRuleset.AsNonNull().KeyBindingInputManager.PressedActions.Contains(action))
```

Comment:
My initial reading of this was "doesn't this mean that if you paused with `LeftButton` held on a slider, and then want to resume holding by clicking the resume cursor indicator with `RightButton`, the next `RightButton` input will get blocked still?" But in testing that... does not appear to be the case for whatever reason? Maybe that has something to do with that entire key up/down syncing/not syncing crap or whatever. Not sure. Any commentary to offer on that?

Reply:
I...think I would want to stick to just block next frame only in that case.

Reply:
I mean I'd also be fine with "if either left or right button is pressed on the ruleset, don't block the new input" or something.

Reply:
Uhhh... can we just not block either gameplay action if either gameplay action was already pressed before break?

Reply:
As in change the conditional to `!PressedActions.Any()`? Well that would be technically wrong.

If you have one action held, you pause, and then resume with the other action. Then, if there's a hit circle at the orange cursor, it'll be hit as a result of you resuming with the other action.

Reply:
It's different for sliders because input is tracked by checking `PressedActions`. This entire input-blocking concept is made specifically for hit circles, it does not apply to sliders and spinners. Because, well, that's how stable does it, apparently?

Reply:
Dear lord.

As long as we've got test coverage of all this it's probably fine to go with... Maybe add a test case covering the resume-with-other-key case.

Reply:
Yep.

Reply:
> It's different for sliders because input is tracked by checking `PressedActions`. This entire input-blocking concept is made specifically for hit circles which receives input by `OnPressed` event instead, it does not apply to sliders and spinners. Because, well, that's how stable does it, apparently?

I don't get this. Does this mean a slider will/"should" start tracking immediately if you start holding a new button before unpausing? Because that sounds wrong to me, ignoring whatever stable may be doing.

Like, as far as I understand what we want is very simple on paper:

- If a key was held before pause, and the key is held after, then it should remain held
- If a key was not held before pause, and is held when unpause occurs, it should not be seen by gameplay

Reply:
I think that for sliders specifically, if you're holding a slider with left key and then want to unpause and continue holding the slider with right key, the input should *not* get dropped, no? Dropping that input just seems anti-user, they'd have to remember which key they paused with before.

Reply:
Maybe it sounds anti-user, but allowing it leads to concerns of using pause buffering as an abuse vector. Also it feels weird that holds are seen but new presses aren't?

Reply:
@peppy There's a difference between what you said here and what I'm conveying here. If the user held a key, then unpaused with another key, only the key that was used for unpausing will be seen by gameplay, the other will not be.

As far as I'm aware, you can change keys while holding a slider, so there's no issue here?

Reply:
The key explicitly used to trigger the unpause can be seen as a hold only, i guess, sure. It's still a bit haphazard (that it only happens for hold not click) but sure.